[ADD] stock_picking_group_by_partner_by_carrier

gurneyalex opened the pull request:
This module can be used if your customers expect that several different orders
they passed will be shipped in a single delivery order.

With this module installed, when a sale order is confirmed, the stock moves for
the lines of the sale order can be placed in an existing delivery order that
shares the same delivery address and carrier (or lack thereof).

Sale orders with a Shipping Policy set to 'When all products are ready' always
get their own shipping.

gurneyalex committed:
[ADD] stock_picking_group_by_partner_by_carrier

This module can be used if your customers expect that several different orders
they passed will be shipped in a single delivery order.

With this module installed, when a sale order is confirmed, the stock moves for
the lines of the sale order can be placed in an existing delivery order that
shares the same delivery address and carrier (or lack thereof).

Sale orders with a Shipping Policy set to 'When all products are ready' always
get their own shipping.

guewen commented on stock_picking_group_by_partner_by_carrier/models/stock_move.py:
```diff
@@ -0,0 +1,78 @@
+                ('picking_type_id', '=', self.picking_type_id.id),
+                ('printed', '=', False),
+                ('immediate_transfer', '=', False),
+                ('state', 'in', states)]
```
Shouldn't you add the filter on `move_type` as well here? Or is the filter on `group_id` enough?

gurneyalex replied:
it is. 
and I have an automated test for it :)

jgrandguillaume reviewed (approved):
Works perfectly !

simahawk commented on stock_picking_group_by_partner_by_carrier/models/stock_picking.py:
```diff
@@ -0,0 +1,99 @@
+        if self.env.context.get("picking_no_overwrite_partner_origin"):
+            written_fields = set(values.keys())
+            if written_fields == {"partner_id", "origin"}:
+                values = {}
```
any chance to land here w/ other values you'd better preserve?

simahawk commented on stock_picking_group_by_partner_by_carrier/models/stock_picking.py:
```diff
@@ -0,0 +1,99 @@
+        ).copy(defaults)
+
+    def do_something(self):
+        return "bla bla"
```
I hope it does :smile:

simahawk commented on stock_picking_group_by_partner_by_carrier/models/stock_picking.py:
```diff
@@ -0,0 +1,99 @@
+            moves = self.move_lines.sorted(lambda m: m.sale_line_id.order_id)
+            if len(moves.mapped("sale_line_id.order_id")) > 1:
+                sales_and_moves = []
+                for sale, sale_moves in groupby(
```
can we split this part as it's the same as at line 55?

florian-dacosta commented:
Thanks for the module
Just one functional remark, I think that not allowing to group "When all products are ready" pickings is too restrictive.
A lot of business configure it this way to avoid partial delivery, which maybe expensive. But it does not mean they would not want to group this kind of picking, too reduce the shipping cost.

I think we should at least have the possibility to group the pickings having this shipping policy if needed, maybe making it configurable or something.
What do you think @gurneyalex ?

gurneyalex committed:
[FIX] cancellation of sale order and WH multi step

* the grouped picking must not be cancelled

* in WH multi step mode:
** ensure the picking ids are computed correctly
** the sale order cancelation does what is expected

gurneyalex commented:
@florian-dacosta Yes we can make this configurable. I'll add the code for this. Thanks for your feedback.

sebalix committed:
[FIX] group pick operations as well with pick-ship delivery steps

jgrandguillaume committed:
Merge pull request #3 from sebalix/13.0-stock_picking_group_by_partner_by_carrier-imp

[FIX] group pick operations as well with pick-ship delivery steps

sebalix committed:
[FIX] stock_picking_group_by_partner_by_carrier: add a hook to handle domain on move_type

And add a test with the 'pick_pack_ship' configuration with grouping
enabled on pack+ship (but not pick).

jgrandguillaume commented:
/ocabot merge

jgrandguillaume commented:
/ocabot merge nobump

guewen commented:
@gurneyalex needs to merge this first https://github.com/gurneyalex/stock-logistics-workflow/pull/4

jgrandguillaume committed:
Merge pull request #4 from sebalix/13.0-stock_picking_group_by_partner_by_carrier-imp2

[FIX] stock_picking_group_by_partner_by_carrier: add a hook to handle domain on move_type

pedrobaeza commented:
ocabot stuck. Please relaunch

jgrandguillaume commented:
/ocabot merge minor

jgrandguillaume commented:
/ocabot merge nobump
